app/models/doi/graphql_query.rb

```diff
@@ -0,0 +1,355 @@
+
+module Doi::GraphqlQuery
+  class Builder
+    include Modelable
+
+    DEFAULT_CURSOR = [0, ""]
+    DEFAULT_PAGE_SIZE = 0
+    DEFAULT_FACET_COUNT = 10
+
+    def initialize(query, options)
+      @query = query
+      @options = options
+    end
+
+    def search_query
+      {
+        size: size,
+        search_after: search_after,
+        sort: sort,
+        query: inner_query,
+        aggregations: aggregations,
+        track_total_hits: true,
+      }.compact
+    end
+
+    def size
+      (@options.dig(:page, :size) || DEFAULT_PAGE_SIZE).to_i
+    end
+
+    def sort
+      [{ created: "asc", uid: "asc" }]
+    end
+
+    def query_fields
+      [
+        "uid^50",
+        "related_identifiers.relatedIdentifier^3",
+        "titles.title^3",
+        "creator_names^3",
+        "creators.id^3",
+        "publisher^3",
+        "descriptions.description^3",
+        "subjects.subject^3"
+      ]
+    end
+
+    def cursor
+      tmp_cursor = @options.dig(:page, :cursor)
+      if tmp_cursor.nil?
+        return DEFAULT_CURSOR
+      end
+
+      if tmp_cursor.is_a?(Array)
+        timestamp, uid = tmp_cursor
+      elsif tmp_cursor.is_a?(String)
+        timestamp, uid = tmp_cursor.split(",")
+      end
+      [timestamp.to_i, uid.to_s]
+    end
+
+    def search_after
+      cursor
+    end
+
+    QUERY_SUBSTITUTIONS = {
+      "publicationYear" => "publication_year",
+      "relatedIdentifiers" => "related_identifiers",
+      "relatedItems" => "related_items",
+      "rightsList" => "rights_list",
+      "fundingReferences" => "funding_references",
+      "geoLocations" => "geo_locations",
+      "landingPage" => "landing_page",
+      "contentUrl" => "content_url",
+      "citationCount" => "citation_count",
+      "viewCount" => "view_count",
+      "downloadCount" => "download_count"
+    }
+
+
+    def clean_query
+      # make sure field name uses underscore
+      # escape forward slash, but not other Elasticsearch special characters
+      if @query.present?
+        @query.gsub(/publicationYear/, "publication_year")\
+          .gsub(/relatedIdentifiers/, "related_identifiers")\
+          .gsub(/relatedItems/, "related_items")\
+          .gsub(/rightsList/, "rights_list")\
+          .gsub(/fundingReferences/, "funding_references")\
+          .gsub(/geoLocations/, "geo_locations")\
+          .gsub(/landingPage/, "landing_page")\
+          .gsub(/contentUrl/, "content_url")\
+          .gsub(/citationCount/, "citation_count")\
+          .gsub(/viewCount/, "view_count")\
+          .gsub(/downloadCount/, "download_count")\
+          .gsub("/", "\\/")
+      else
+        ""
+      end
+    end
+
+    def must
+      if !@query.present?
+        [{ match_all: {} }]
+      else
+        [{
+          query_string: {
+            query: clean_query,
+            fields: query_fields,
+            default_operator: "AND",
+            phrase_slop: 1
+          }
+        }]
+      end
+    end
+
+    def filters
```

Comment:
Part of me wonders if each filter should be a seperate function.
e.g.
```ruby
def add_filter_provider_id
[{ terms: { provider_id: options[:provider_id].split(",") } } ]
```

So then you'd do the options check where you combine all the filters. This could have the benefit that you could build a query via composing the functions, if that makes sense?

Reply:
I have some ideas for that (Specifically around aggregations).  For this refactor I wanted to keep things as close as the original functionality.

I think the next steps would be to add more tests for filters, and then look at refactoring the filters.